### PR TITLE
Option to skip files and directories and omit isort or black

### PR DESCRIPTION
**Pull request by ismet55555**
Added a `--skip` option for pyfmt to skip files and/or directories
Useful for when you have to skip the virtual environment (`env`, `venv`, etc)  directory or something like that.

Example for skipping directories. Will skip everything in the directory.
It must be relative to current directory format. 
- `--skip=dir/subdir`

Example for specific files. Will skip specific files.
- `--skip=yo.py,mama.py`

Example for mixing things up with files and directories:
- `--skip=daddy.py,dir/subdir`

If it can't find a file or directory it will stop pyfmt with a error message. It would be pretty bad to want to skip a file, misspell it, and it formats it anyways!

Basically:
- isort: --skip
- black: --exclude

=====================================================

Added the ability to only run `isort` or `black`, not both.
I also changed some minor print statements.

This is based on the branch form previous work/branch

Sorry for the crappy programming!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,11 +55,16 @@ positional arguments:
 
 optional arguments:
   -h, --help            show this help message and exit
+  --skip                Directory (ie. env/foo) or files (ie. cool.py) to skip
+                        Arguments are seperated by comma (ie. env,kung.py,foo.py)
+  --isort               Only process with isort (imports only)
+  --black               Only process with black (code only)
```
Comment (LyleScott) [inline README.md]: these names can be more descriptive to their purpose

Something like 
`--isort-only`
`--black-only`

File: README.md [hunk 2]
```diff
@@ -54,11 +55,16 @@ positional arguments:
...
   --check               don't write changes, just print the files that would
                         be formatted
   --line-length LINE_LENGTH
                         max characters per line; defaults to $MAX_LINE_LENGTH
                         or 100
+  --show-title          show a fancy title at the beginning of pyfmt
```
Comment (LyleScott) [inline README.md]: while neat, i don't think anyone would desire this to be on and i'd rather not add superfluous code to the codebase

anyone have an opinion? i'm a fan of less code in general...

Comment (LyleScott) [reply README.md]: it's also using bash specific escape codes and not all users are using a bash (or derivative)

File: README.md [hunk 3]
```diff
@@ -43,7 +43,8 @@ If you are not in your virtual env, the global Python environment will be used w
 your local package imports in with the 3rd party package imports.
 
 ```console
-usage: pyfmt [-h] [--check] [--line-length LINE_LENGTH]
+usage: pyfmt [-h] [--skip] [--isort] [--black] [--check] [--line-length LINE_LENGTH]
```
Comment (LyleScott) [inline README.md]: let's get the kwargs on their own line, alphabetized.

args can still live on that top line, but alphabetize

**Comment by LyleScott** [comment]
examples of skipping should be in the README i think

File: pyfmt/__init__.py [hunk 4]
```diff
@@ -25,19 +28,139 @@
...
+        )
+
+    # Executing black code formatter
+    black_exitcode = 0
+    if isort_black_run[1]:
+        black_exitcode = run_formatter(
+            BLACK_CMD, path, line_length=line_length, extra_black_args=extra_black_args
+        )
 
+    display_divider(title="pyfmt Execution Time: {0:.2f} sec".format(time.time() - timer_start))
```
Comment (LyleScott) [inline pyfmt/__init__.py]: why does execution time matter? aren't almost all things subsecond? Do these stats mean anything to anyone? I'd rather not add stuff that doesn't have a value.

File: pyfmt/__init__.py [hunk 5]
```diff
@@ -25,19 +28,139 @@
...
+            elif os.path.abspath(item) in all_dirs:
+                # Saving all filenames in user specified directory
+                files_in_dir = []
+                for (dirpath, dirnames, filenames) in os.walk(item):
+                    for filename in filenames:
+                        if filename.split(".")[-1] in ("py", "pyi"):
+                            files_in_dir.append(filename)
+                filenames_to_skip.extend(files_in_dir)
+            else:
+                print(f'CRITICAL: One of the files or directories marked as skipped not found ("{item}").')
```
Comment (LyleScott) [inline pyfmt/__init__.py]: if the user tried to skip it and it wasn't found, why is this a terminal condition? In either case, it won't run which is the desired behavior?

File: pyfmt/__init__.py [hunk 6]
```diff
@@ -25,19 +28,139 @@
...
+                sys.exit()
+        # Display Files skipped
+        display_divider(title="SKIPPING FILES")
+        for filename_to_skip in filenames_to_skip:
+            print(f"SKIPPING: {filename_to_skip}")
+        print(f"\nNumber of files to be skipped by pyfmt: {len(filenames_to_skip)}")
+        # Make a continuos string of arguments for
+        #   isort - must be separate --skip for each file
+        #   black - regex for exact filename (ie. file1|file2|etc.)
+        isort_filenames_to_skip = "--skip=" + " --skip=".join(filenames_to_skip)
```
Comment (LyleScott) [inline pyfmt/__init__.py:110]: this would result in a line like
```
--skip= --skip=foo --skip=bar
```

it seems like the first naked `--skip=` is unneeded, right?

is that

File: pyfmt/__init__.py [hunk 7]
```diff
@@ -25,19 +28,139 @@
...
+            else:
+                print(f'CRITICAL: One of the files or directories marked as skipped not found ("{item}").')
+                print("CRITICAL: Check spelling or existence of file or directory")
+                print("CRITICAL: Aborting pyfmt ...")
+                sys.exit()
+        # Display Files skipped
+        display_divider(title="SKIPPING FILES")
+        for filename_to_skip in filenames_to_skip:
+            print(f"SKIPPING: {filename_to_skip}")
+        print(f"\nNumber of files to be skipped by pyfmt: {len(filenames_to_skip)}")
```
Comment (LyleScott) [inline pyfmt/__init__.py]: i'm not sure this status line means anything to anyone.. perhaps we can remove it

File: pyfmt/__init__.py [hunk 8]
```diff
@@ -25,19 +28,139 @@
...
+        # Make a continuos string of arguments for
+        #   isort - must be separate --skip for each file
+        #   black - regex for exact filename (ie. file1|file2|etc.)
+        isort_filenames_to_skip = "--skip=" + " --skip=".join(filenames_to_skip)
+        black_filenames_to_skip = "--exclude=" + "|".join(filenames_to_skip)
+        # Adding to the isort and black arguments
+        extra_isort_args += isort_filenames_to_skip
+        extra_black_args += "--exclude=" + black_filenames_to_skip
```
Comment (LyleScott) [inline pyfmt/__init__.py]: this would result in a variable value `--exclude=--exclude=foo|bar|baz` it looks like. seems like a bug?

File: pyfmt/__init__.py [hunk 9]
```diff
@@ -25,19 +28,139 @@
...
         extra_black_args += " --check"
 
-    isort_exitcode = run_formatter(
-        ISORT_CMD, path, line_length=line_length, extra_isort_args=extra_isort_args
-    )
-    black_exitcode = run_formatter(
-        BLACK_CMD, path, line_length=line_length, extra_black_args=extra_black_args
-    )
+    # If user specified to run only isort or black
+    isort_black_run = [True, True]
```
Comment (LyleScott) [inline pyfmt/__init__.py]: this is unneeded. you already have bools for these flags in variables.

File: pyfmt/__init__.py [hunk 10]
```diff
@@ -25,19 +28,139 @@
...
+    if isort:
+        isort_black_run = [True, False]
+    if black:
+        isort_black_run = [False, True]
+
+    # Executing isort import formatter
+    isort_exitcode = 0
+    if isort_black_run[0]:
```
Comment (LyleScott) [inline pyfmt/__init__.py]: why not use `isort`?  (should be `isort_only`)

same for below, why not just use `black` (should be `black_only`)?

File: README.md (hunk 2 above)
Comment (augustawind) [reply README.md]: yeah, I'd rather not have this in here at all

File: pyfmt/__init__.py (hunk 6 above)
Comment (augustawind) [reply pyfmt/__init__.py:110]: `join` inserts between each item, not before, so i'm pretty sure this would create the correct output.

File: pyfmt/__init__.py (hunk 5 above)
Comment (ismet55555) [reply pyfmt/__init__.py]: it's a terminal condition because if the file was not found, more than likely the user made a spelling mistake, in which case the file they wanted to exclude got formatted.  I might be thinking about this wrong but I would figure I don't want specific files to be formatted by mistake...

File: pyfmt/__init__.py (hunk 7 above)
Comment (ismet55555) [reply pyfmt/__init__.py]: It's a subjective thing of course, but it's nice to have.   The black module prints out emojis at the end, which seems on one of the spectra of nice to have or charming, which doesn't really mean anything to anyone

If it's completely useless, I can remove this and the timing.  But to me it seems like a reasonable piece of information especially if you are omitting an entire directory for this, and for the timing if someone quickly needs a reference in performance.

File: pyfmt/__init__.py (hunk 8 above)
Comment (ismet55555) [reply pyfmt/__init__.py]: good catch!.... not sure why I changed the `isort` part and not the `black` part

File: pyfmt/__init__.py (hunk 9 above)
Comment (ismet55555) [reply pyfmt/__init__.py]: yeah i was trying to think of if there are more formatting modules available... so if you only want 1/5 modules or 2/5 modules to run.   

I could do something like `[isort, black]` and if there are more modules `[isort, black, module3, module4]` ... probably rename the variable `isort_black_run` to something like `modules_run` .... probably make it a dict `{isort: bool, black: bool, module3: bool, etc }`

File: README.md (hunk 1 above)
Comment (ismet55555) [reply README.md]: that's no problem, thought about it, but in the context of pyfmt what else could it mean to just have `isort` instead of `isort-only`

I'll change it

File: README.md (hunk 2 above)
Comment (ismet55555) [reply README.md]: yeah i have no problem with removing it, ... oh well, there was an attempt haha

File: pyfmt/__init__.py (hunk 7 above)
Comment (augustawind) [reply pyfmt/__init__.py]: I'm with lyle... let's get the output lean n mean. I think what we had before was fine 🤷‍♀

Comment (ismet55555) [reply pyfmt/__init__.py]: cool, peer pressure wins haha ... I'll remove it

File: pyfmt/__init__.py (hunk 4 above)
Comment (ismet55555) [reply pyfmt/__init__.py]: that depends on the project, of course, with system test it can take a couple seconds
but again, meh, I'll remove it

File: .gitignore [hunk 11]
```diff
@@ -1,5 +1,9 @@
 .vscode
+.idea/
 /build/
 /dist/
 *.egg-info
 __pycache__
+env/
+venv/
+*.pyc
```
Comment (shawnjoseph) [inline .gitignore:9]: looks like github wants you to add a newline, nbd

File: README.md [hunk 12]
```diff
@@ -72,3 +78,7 @@ optional arguments:
   you run the code will be formatted.
 * As a matter of convenience, you may set `$BASE_CODE_DIR` in your environment and run the script
   in the case this is better for your needs.
+* `--skip` directory names must be relative to current directory. For example, if the current directory
+  contains the directory `env/bin`, this entire directory can be skipped with `--skip=env/bin`
+* `--skip` file names can not include the directory names, only file names are allowed. For example, 
+  if there is file name "foo/bar/yo.py", it is skipped by `--skip=yo.py`
```
Comment (shawnjoseph) [inline README.md:85]: I'm assuming this also means that we can only skip files in the current directory?

Comment (shawnjoseph) [reply README.md:85]: ah maybe not, reading lower it seems all files within the dir/sub_dirs are fair game

File: pyfmt/__init__.py [hunk 13]
```diff
@@ -24,19 +27,101 @@
...
+            else:
+                print(f'CRITICAL: One of the files or directories marked as skipped not found ("{item}").')
+                print("CRITICAL: Check spelling or existence of file or directory")
+                print("CRITICAL: Aborting pyfmt ...")
+                sys.exit()
+        # Display Files skipped
+        display_divider(title="SKIPPING FILES")
+        for filename_to_skip in filenames_to_skip:
+            print(f"SKIPPING: {filename_to_skip}")
+        # Make a continuos string of arguments for
```
Comment (shawnjoseph) [inline pyfmt/__init__.py]: continuous* 🤷‍♂

**Comment by shawnjoseph** [comment]
lgtm. Did you run pyfmt on this? 😄

**Comment by LyleScott** [comment]
This looks great! Thanks.

I'm going to check this out and give it a go for a little QA.

**Comment by LyleScott** [comment]
@ismet55555 `pyfmt` seems to no longer do anything unless you supply `--isort-only` or `--black-only` arguments. The default case, and current behavior, should be running `isort` _and_ `black` if the user did not supply an _only_ argument. 

To repeat:

```
# These work fine:
python3 -m pyfmt --isort-only
python3 -m pyfmt --black-only

# This now does nothing when it should do everything
python3 -m pyfmt
```

**Review by LyleScott — CHANGES_REQUESTED** [review]
@ismet55555 pyfmt seems to no longer do anything unless you supply --isort-only or --black-only arguments. The default case, and current behavior, should be running isort and black if the user did not supply an only argument.

To repeat:

```
# These work fine:
python3 -m pyfmt --isort-only
python3 -m pyfmt --black-only

# This now does nothing when it should do everything
python3 -m pyfmt
```

**Comment by LyleScott** [comment]
LGTM, just update paths to skipped files to be full relative [or absolute?] paths instead of just the filename.
